Installation issue:  hip@6.0.0 fails to build

### Steps to reproduce the issue

@haampie  @renjithravindrankannath  @srekolam
In an attempt to fix the [e4s pipeline error](https://gitlab.spack.io/spack/spack/-/pipelines/610941), I spun up the same image and a reduced spack.yaml to isolate the issue causing specs. After cloning the commit for #42270, and using the spack.yaml below, HIP failed to build, as it did in the pipeline.

```console
$ spack spec -I hip
--------------------------------
 -   hip

Concretized
--------------------------------
 -   hip@6.0.0%gcc@11.4.0~cuda~ipo+rocm build_system=cmake build_type=Release generator=make patches=1f65dfe,3f783ae,aee7249,d699165 arch=linux-ubuntu20.04-x86_64_v3
[e]      ^cmake@3.26.3%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64_v3
[+]      ^comgr@6.0.0%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
[+]          ^ncurses@6.3%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]          ^rocm-cmake@6.0.0%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
[+]          ^z3@4.12.4%gcc@11.4.0~gmp~ipo~python build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
[+]          ^zlib-ng@2.1.5%gcc@11.4.0+compat+opt build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64_v3
 -       ^glx@1.4%gcc@11.4.0 build_system=bundle arch=linux-ubuntu20.04-x86_64_v3
 -           ^mesa@23.2.1%gcc@11.4.0+glx+llvm+opengl~opengles+osmesa~strip build_system=meson buildtype=release default_library=shared arch=linux-ubuntu20.04-x86_64_v3
[+]              ^bison@3.8.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[e]                  ^diffutils@3.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^expat@2.5.0%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libbsd@0.11.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                      ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^flex@2.6.3%gcc@11.4.0+lex~nls build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[e]              ^gettext@0.19.8.1%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools patches=9acdb4e arch=linux-ubuntu20.04-x86_64_v3
[+]              ^glproto@1.4.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libunwind@1.6.2%gcc@11.4.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs+pic+tests+weak_backtrace+xz~zlib build_system=autotools components=none libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libx11@1.8.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^inputproto@2.3.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^kbproto@1.0.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^xextproto@7.3.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^xproto@7.0.31%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^xtrans@1.5.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libxcb@1.16%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libxau@1.0.11%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libxdmcp@1.1.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^xcb-proto@1.16.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libxext@1.3.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libxt@1.3.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libice@1.1.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libsm@1.2.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                      ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
 -               ^llvm@17.0.4%gcc@11.4.0+clang~cuda~flang+gold~ipo+libomptarget~libomptarget_debug~link_llvm_dylib+lld+lldb+llvm_dylib+lua~mlir+polly~python~split_dwarf~z3~zstd build_system=cmake build_type=Release compiler-rt=runtime generator=ninja libcxx=runtime libunwind=runtime openmp=runtime shlib_symbol_version=none targets=all version_suffix=none arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^binutils@2.41%gcc@11.4.0~gas+gold~gprofng+headers~interwork+ld~libiberty~lto~nls~pgo+plugins build_system=autotools compress_debug_sections=zlib libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^hwloc@2.9.1%gcc@11.4.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libedit@3.1-20210216%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libffi@3.4.4%gcc@11.4.0 build_system=autotools patches=070b1f3 arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^lua@5.3.6%gcc@11.4.0~pcfile+shared build_system=makefile fetcher=curl arch=linux-ubuntu20.04-x86_64_v3
[+]                      ^curl@8.4.0%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-x86_64_v3
[+]                          ^nghttp2@1.57.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                          ^openssl@3.1.3%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64_v3
[+]                              ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]                      ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64_v3
[+]                      ^unzip@6.0%gcc@11.4.0 build_system=makefile arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^perl-data-dumper@2.173%gcc@11.4.0 build_system=perl arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^swig@4.1.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                      ^pcre2@10.42%gcc@11.4.0~jit+multibyte build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
 -               ^meson@1.2.2%gcc@11.4.0 build_system=python_pip patches=0f0b1bd,ae59765 arch=linux-ubuntu20.04-x86_64_v3
 -               ^py-mako@1.2.4%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64_v3
 -                   ^py-markupsafe@2.1.3%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64_v3
[+]              ^xrandr@1.5.2%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libxrandr@1.5.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                      ^renderproto@0.11.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^libxrender@0.9.11%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]                  ^randrproto@1.5.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[e]      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]      ^hipify-clang@6.0.0%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make patches=54b8b39 arch=linux-ubuntu20.04-x86_64_v3
 -       ^hsa-rocr-dev@6.0.0%gcc@11.4.0~asan+image~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
[+]          ^elfutils@0.190%gcc@11.4.0~debuginfod+exeprefix~nls build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]              ^xz@5.4.1%gcc@11.4.0+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]              ^zstd@1.5.5%gcc@11.4.0~programs build_system=makefile libs=shared,static arch=linux-ubuntu20.04-x86_64_v3
[+]          ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]          ^xxd-standalone@8.2.1201%gcc@11.4.0 build_system=makefile arch=linux-ubuntu20.04-x86_64_v3
 -       ^hsakmt-roct@6.0.0%gcc@11.4.0~asan~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
 -           ^libdrm@2.4.115%gcc@11.4.0~docs build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]              ^libpthread-stubs@0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]      ^llvm-amdgpu@6.0.0%gcc@11.4.0~ipo~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_system=cmake build_type=Release generator=ninja patches=53f9500 arch=linux-ubuntu20.04-x86_64_v3
[e]          ^ninja@1.11.1%gcc@11.4.0+re2c build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[e]          ^python@3.8.10%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma+nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,4c24573,ebdca64,f2fd060 arch=linux-ubuntu20.04-x86_64_v3
[+]      ^numactl@2.0.14%gcc@11.4.0 build_system=autotools patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu20.04-x86_64_v3
[e]          ^autoconf@2.69%gcc@11.4.0 build_system=autotools patches=7793209 arch=linux-ubuntu20.04-x86_64_v3
[e]          ^automake@1.16.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[+]          ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[e]              ^findutils@4.7.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu20.04-x86_64_v3
[e]          ^m4@1.4.18%gcc@11.4.0+sigsegv build_system=autotools patches=3877ab5,fc9b616 arch=linux-ubuntu20.04-x86_64_v3
[e]      ^perl@5.30.0%gcc@11.4.0~cpanm+opcode+open+shared+threads build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]      ^perl-file-which@1.27%gcc@11.4.0 build_system=perl arch=linux-ubuntu20.04-x86_64_v3
[+]      ^perl-uri-encode@1.1.1%gcc@11.4.0 build_system=perl arch=linux-ubuntu20.04-x86_64_v3
[+]          ^perl-module-build@0.4232%gcc@11.4.0 build_system=perl arch=linux-ubuntu20.04-x86_64_v3
 -       ^py-cppheaderparser@2.7.4%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64_v3
 -           ^py-pip@23.1.2%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64_v3
 -           ^py-ply@3.11%gcc@11.4.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64_v3
 -           ^py-setuptools@68.0.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64_v3
 -           ^py-wheel@0.41.2%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64_v3
[+]      ^rocm-core@6.0.0%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
 -       ^rocminfo@6.0.0%gcc@11.4.0~ipo build_system=cmake build_type=Release generator=make arch=linux-ubuntu20.04-x86_64_v3
[+]      ^roctracer-dev-api@6.0.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu20.04-x86_64_v3

```
Included as well is the minimal spack.yaml I used for this.
[hip-spack.yaml.txt](https://github.com/spack/spack/files/14119315/hip-spack.yaml.txt)


### Error message

<details><summary>HIP fails to build.</summary>

<pre>
  >> 119     /tmp/root/spack-stage/spack-stage-hip-6.0.0-ndctizwtjldnbakn5kv23xcumdopwnkn/spack-src/include/hip/hip_r
             untime.h:41:2: error: HIP is not supported on the specified GPU ARCH with wavefront size 64
     120     #error HIP is not supported on the specified GPU ARCH with wavefront size 64
     121      ^
     122     1 error generated when compiling for gfx803.
  >> 123     make[2]: *** [hipamd/src/hiprtc/CMakeFiles/hiprtc-builtins.dir/build.make:84: hipamd/src/hiprtc/hip_rtc_
             gen/hipRTC] Error 1
     124     make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-hip-6.0.0-ndctizwtjldnbakn5kv23xcumdopwnkn
             /spack-build-ndctizw'
  >> 125     make[1]: *** [CMakeFiles/Makefile2:414: hipamd/src/hiprtc/CMakeFiles/hiprtc-builtins.dir/all] Error 2
     126     make[1]: *** Waiting for unfinished jobs....
....
     1061    [ 57%] Built target rocclr
     1062    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-hip-6.0.0-ndctizwtjldnbakn5kv23xcumdopwnkn
             /spack-build-ndctizw'
  >> 1063    make: *** [Makefile:159: all] Error 2
</pre></details>


### Information on your system

This was done repeated on both the [ci pipeline](https://gitlab.spack.io/spack/spack/-/pipelines/610941) for #42270 , and on a different machine by me, both using the same image, ghcr.io/spack/ubuntu20.04-runner-amd64-gcc-11.4:2023.08.01 
```
$ spack debug report
* **Spack:** 0.22.0.dev0 (2bcfb72d38daa2d148b5551c21a424633339649c)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo
```


### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/14119452/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/14119454/spack-build-out.txt)

Using a cut down version of the e4s/spack.yaml that the pipeline used, containing only the root specs:
```
  specs:
  - py-tensorflow ~cuda +rocm amdgpu_target=gfx908
  - py-tensorflow ~cuda +rocm amdgpu_target=gfx90a
```
I was able to get the same error as the pipeline. There are a few things that stand out to me, 
1. It says failed to compile for gfx803, but there is no gfx803 on the system.
2. In the build out txt, on line 44, the compile command has the two flags `--cuda-device-only --cuda-gpu-arch=gfx1030`, but the gfx1030 is not a  cuda card.

My thoughts are that somewhere on the system HIP is supposed to either be handed the card to compile for or it is supposed to detect  it through rocminfo??? and somewhere this process is failing, which is why it says it is failing to compile for gfx803, rather than the gfx1030, or the gfx908 as requested by the root spec.

I found [here](https://rocm.docs.amd.com/projects/HIP/en/latest/doxygen/html/hip__runtime_8h_source.html), the source file where the error is coming from. 



### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/hip/package.py
+++ b/var/spack/repos/builtin/packages/hip/package.py
@@ -531,12 +531,6 @@ class Hip(CMakePackage):
                 "clr/hipamd/hip-config-amd.cmake",
                 string=True,
             )
-            filter_file(
-                '"${ROCM_PATH}/llvm"',
-                self.spec["llvm-amdgpu"].prefix,
-                "clr/hipamd/src/hiprtc/CMakeLists.txt",
-                string=True,
-            )
         perl = self.spec["perl"].command
 
         if self.spec.satisfies("@:5.5"):
@@ -561,7 +555,12 @@ class Hip(CMakePackage):
                     filter_file(" -lnuma", f" -L{numactl} -lnuma", "hipBin_amd.h")
 
     def cmake_args(self):
-        args = []
+        args = [
+            # find_package(Clang) and find_package(LLVM) in clr/hipamd/src/hiprtc/CMakeLists.txt
+            # should find llvm-amdgpu
+            self.define("LLVM_ROOT", self.spec["llvm-amdgpu"].prefix),
+            self.define("Clang_ROOT", self.spec["llvm-amdgpu"].prefix),
+        ]
         if self.spec.satisfies("+rocm"):
             args.append(self.define("HSA_PATH", self.spec["hsa-rocr-dev"].prefix))
             args.append(self.define("HIP_COMPILER", "clang"))

